Enhanced API and `FlipTransition`

m-haisham opened the pull request:
# Description

This is a major change to the codebase and includes api changes.

## Added

- Added `flutter_lints` to the library
- Added `FlipTransition` widget
- Added target side support for flip and flipWithoutAnimation

## Changed

- Replaced `FlipDirection` with `Axis`
- Replace speed with duration on `FlipCard`
- Rewrote hint method to be more simple to use
- Remove isFront status and rely on animation status
- Rename toggleCard method to flip
- Moved skew and hint to `FlipCardState`
## Fixed

- Unassign state from controller on dispose
- Unassociate old controller when replaced
 
## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

- [x] Self-Code Review
- [x] Tested on example project

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Changed the example project to contain the new feature (only if applies)

Commit:
Add flutter lints to library

Commit:
Replace FlipDirection with Axis

Commit:
Multiple changes from flutter lint

Commit:
Remove fill prefix from Fill enum variants

Commit:
Make AnimationController non-nullable

Commit:
Replace speed with duration

Commit:
Minor refactoring on animation card

Commit:
Unassign state from controller on dispose

Commit:
Extracted flip card transition widget

Commit:
Changed hint to rely on skew percentage instead of duration

Commit:
Remove isFront state and rely on animate controller status

Commit:
Fix toggle card without animation

Commit:
Rename toggle card to flip

Commit:
Hide already showing snackbar before showing next

Commit:
Fix and test ignore pointer change

Commit:
Different color for back side of the card

Commit:
Move control methods to flip card state

Commit:
Rename flip_card_transition to flip_transition

Commit:
Detach controller from state when disposed

Commit:
Added more documentation

Commit:
Unassociate old controller when replaced

Commit:
Update README.md

Commit:
Bump minimum dart version to 2.17

Commit:
Line artifact fixes

Commit:
Add flip target side support

Commit:
Add target side support to controller

Commit:
Minor fixes and documentation

ciriousjoker commented on README.md:
```diff
@@ -1,39 +1,87 @@
+Create a flip card as shown below. By default the card is touch controlled.
 
-Create a flip card. The card will flip when touched
+You can turn of touch control by setting `flipOnTouch` to `false`.
```
typo "off"

ciriousjoker commented on lib/flip_transition.dart:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  CardSide _getSideFor(AnimationStatus status) {
+    switch (status) {
```
This seems dangerous, example:
1. user taps to flip
2. flip animation starts (`AnimationStatus.forward`)
3. app uses the _currentSide property (returns `CardSide.back` even though the front of the card is still visible)

As far as I can tell, the _currentSide property is only used internally so this should be fine. Any thoughts?

m-haisham replied:
If the user taps the card before the animation is completed, the animation will be reversed to the other side. In this case, when animating towards the back side of the card, it will start animating towards the front again.

My thought process was if a user wants to cancel the animation before it is completed, it is possible this way.

ciriousjoker replied:
Got it, makes sense

ciriousjoker commented on lib/flip_transition.dart:
```diff
@@ -0,0 +1,182 @@
+  @override
+  Widget build(BuildContext context) {
+    final transform = Matrix4.identity();
+    transform.setEntry(3, 2, 0.001);
```
Where do these magic numbers come from? Why exactly `0.001`? Is any small number ok? Why not exactly 0? Why not `double.minPositive`? Please add comment.

m-haisham replied:
The number was already present in the code, I tried it with a couple of numbers including `double.minPositive` but it doesn't give the desired effect.

ciriousjoker commented on pubspec.yaml:
```diff
@@ -4,13 +4,14 @@ version: 0.7.0
     sdk: flutter
 
 dev_dependencies:
+  flutter_lints: ^2.0.1
```
I personally prefer the [lint](https://pub.dev/packages/lint) package as it's pretty strict. That being said, any linter is a big improvement already :D

m-haisham replied:
I have personally never tried lint package but, I use flutter_lints because its the recommended set of lints by flutter team.

Should I switch?

ciriousjoker replied:
No it's ok, leave it as is.

ciriousjoker commented on lib/flip_card.dart:
```diff
@@ -1,255 +1,303 @@
-    });
+    switch (targetSide) {
+      case CardSide.front:
+        await controller.forward().complete;
```
The flip() animation doesn't work at all.
However, it does work once I switch `forward()` with `reverse()` (line 174).
Maybe it's better to fix this in a different way though.

m-haisham replied:
I've fixed it

ciriousjoker reviewed (changes requested):
@m-haisham Sorry for taking so long, here's the review:

- Code is very well written in general
- Nice usage of `@template` & `@macro`. 
- Functional "style" (e.g. `_getSideFor`) without using it dogmatically
- `extends AnimatedWidget` is a great idea

Sadly, the flip animation doesn’t work. I've suggested a solution in one of the comments.

Commit:
Update gradle clean task syntax

Commit:
Fix flip card flip and flipWithoutAnimation method

Commit:
Move card side related method to enum

Commit:
Rename flip_transition to flip_card_transition and animation_card to flip_transition

Now FlipTransition is a pure transition, unlike previously where there where two transitions inside it

Commit:
Fix typo from "of" to "off"

ciriousjoker reviewed (approved):
lgtm. @BrunoJurkovic this is a breaking change and since we're below 1.0.0 so far, bumping to v0.8.0 should be enough.
